File: itx-go/itx-impl/queue/rabbitmq/factory.go
package rabbitmq

import (
	"fmt"
	"os"
	"strconv"

	"github.com/chehsunliu/itx/itx-go/itx-contract/queue"
	amqp "github.com/rabbitmq/amqp091-go"
)

const defaultMaxConcurrency = 100

type MessageQueueFactory struct {
	conn                     *amqp.Connection
	maxConcurrency           int64
	controlStandardQueueName string
	controlPremiumQueueName  string
	computeStandardQueueName string
	computePremiumQueueName  string
}

func FromEnv() (*MessageQueueFactory, error) {
	host := os.Getenv("ITX_RABBITMQ_HOST")
	port := os.Getenv("ITX_RABBITMQ_PORT")
	user := os.Getenv("ITX_RABBITMQ_USER")
	password := os.Getenv("ITX_RABBITMQ_PASSWORD")

	url := fmt.Sprintf("amqp://%s:%s@%s:%s/%%2F", user, password, host, port)
	conn, err := amqp.Dial(url)
	if err != nil {
		return nil, err
	}

	maxConcurrency := int64(defaultMaxConcurrency)
	if raw := os.Getenv("ITX_RABBITMQ_MAX_CONCURRENCY"); raw != "" {
		if v, err := strconv.ParseInt(raw, 10, 64); err == nil && v > 0 {
			maxConcurrency = v
		}
	}

	return &MessageQueueFactory{
		conn:                     conn,
		maxConcurrency:           maxConcurrency,
		controlStandardQueueName: os.Getenv("ITX_RABBITMQ_CONTROL_STANDARD_QUEUE"),
		controlPremiumQueueName:  os.Getenv("ITX_RABBITMQ_CONTROL_PREMIUM_QUEUE"),
		computeStandardQueueName: os.Getenv("ITX_RABBITMQ_COMPUTE_STANDARD_QUEUE"),
		computePremiumQueueName:  os.Getenv("ITX_RABBITMQ_COMPUTE_PREMIUM_QUEUE"),
	}, nil
}

func (f *MessageQueueFactory) CreateControlStandardQueue() queue.MessageQueue {
	return newMessageQueue(f.conn, f.controlStandardQueueName, f.maxConcurrency)
}

func (f *MessageQueueFactory) CreateControlPremiumQueue() queue.MessageQueue {
	return newMessageQueue(f.conn, f.controlPremiumQueueName, f.maxConcurrency)
}

func (f *MessageQueueFactory) CreateComputeStandardQueue() queue.MessageQueue {
	return newMessageQueue(f.conn, f.computeStandardQueueName, f.maxConcurrency)
}

func (f *MessageQueueFactory) CreateComputePremiumQueue() queue.MessageQueue {
	return newMessageQueue(f.conn, f.computePremiumQueueName, f.maxConcurrency)
}

File: itx-rs/itx-impl/src/queue/rabbitmq/factory.rs
use std::sync::Arc;

use itx_contract::queue::MessageQueue;
use itx_contract::queue::factory::MessageQueueFactory;
use lapin::{Connection, ConnectionProperties};

use crate::queue::rabbitmq::RabbitMessageQueue;

#[derive(serde::Deserialize)]
struct RabbitMessageQueueFactoryConfig {
    pub host: String,
    pub port: u16,
    pub user: String,
    pub password: String,
    #[serde(default = "default_max_concurrency")]
    pub max_concurrency: u32,
    pub control_standard_queue: String,
    pub control_premium_queue: String,
    pub compute_standard_queue: String,
    pub compute_premium_queue: String,
}

fn default_max_concurrency() -> u32 {
    100
}

pub struct RabbitMessageQueueFactory {
    conn: Arc<Connection>,
    config: RabbitMessageQueueFactoryConfig,
}

impl RabbitMessageQueueFactory {
    pub async fn from_env() -> Result<Self, lapin::Error> {
        let config = envy::prefixed("ITX_RABBITMQ_")
            .from_env::<RabbitMessageQueueFactoryConfig>()
            .expect("failed to read RabbitMQ environment variables");

        let url = format!(
            "amqp://{}:{}@{}:{}/%2F",
            config.user, config.password, config.host, config.port
        );
        let conn = Connection::connect(&url, ConnectionProperties::default()).await?;

        Ok(Self {
            conn: Arc::new(conn),
            config,
        })
    }
}

impl MessageQueueFactory for RabbitMessageQueueFactory {
    fn create_control_standard_queue(&self) -> Arc<dyn MessageQueue> {
        Arc::new(RabbitMessageQueue::new(
            self.conn.clone(),
            self.config.control_standard_queue.clone(),
            self.config.max_concurrency,
        ))
    }

    fn create_control_premium_queue(&self) -> Arc<dyn MessageQueue> {
        Arc::new(RabbitMessageQueue::new(
            self.conn.clone(),
            self.config.control_premium_queue.clone(),
            self.config.max_concurrency,
        ))
    }

    fn create_compute_standard_queue(&self) -> Arc<dyn MessageQueue> {
        Arc::new(RabbitMessageQueue::new(
            self.conn.clone(),
            self.config.compute_standard_queue.clone(),
            self.config.max_concurrency,
        ))
    }

    fn create_compute_premium_queue(&self) -> Arc<dyn MessageQueue> {
        Arc::new(RabbitMessageQueue::new(
            self.conn.clone(),
            self.config.compute_premium_queue.clone(),
            self.config.max_concurrency,
        ))
    }
}
